components/raftstore/src/store/peer.rs

```diff
@@ -4414,6 +4425,33 @@ where
         pessimistic_locks.is_valid = false; // Not necessary, but just make it safer.
         pessimistic_locks.map = Default::default();
     }
+
+    pub fn need_renew_lease_at<T>(
+        &self,
+        ctx: &PollContext<EK, ER, T>,
+        renew_bound: Timespec,
+    ) -> bool {
+        let max_lease = ctx.cfg.raft_store_max_leader_lease();
+        let has_overlapped_reads = self.pending_reads.back().map_or(false, |read| {
+            // If there is any read index whose lease can cover till next heartbeat
+            // then we don't need to propose a new one
+            read.propose_time + max_lease > renew_bound
+        });
+        let has_overlapped_writes = self.proposals.back().map_or(false, |proposal| {
+            // If there is any write whose lease can cover till next heartbeat
+            // then we don't need to propose a new one
+            proposal
+                .propose_time
+                .map_or(false, |propose_time| propose_time + max_lease > renew_bound)
+        });
+        if has_overlapped_reads || has_overlapped_writes {
+            return false;
+        }
+        matches!(
+            self.leader_lease.inspect(Some(renew_bound)),
+            LeaseState::Expired
+        )
```

Comment:
`LeaseState::Valid` should be a common case, how about moving this to the start of the function to make a fast path?

Reply:
Good idea, fixed.

Reply:
Why do you think it's a fast path? The lease is extended due to being woken up. So there probably are pending requests already that can extend the lease without this action. So by putting the query at the end of the function, access to monotonic raw time is reduced for all hot regions and all cold regions until it's at the end of its lease.

Reply:
Actually there is no access to monotonic raw time (we have got a monotonic raw time as a parameter of this function), and the interval of this tick was set to `max_lease_time / 4`, so there probably is a big chance that the lease is not expired here.🤔

Reply:
Got it.